Source/Initialization/InitSpaceChargeField.cpp

```diff
@@ -251,34 +259,44 @@ WarpX::computeB (amrex::Vector<std::array<std::unique_ptr<amrex::MultiFab>, 3> >
+                        -beta_y*inv_dz*0.5*(phi_arr(i,j  ,k+1)-phi_arr(i,j  ,k)
+                                          + phi_arr(i,j+1,k+1)-phi_arr(i,j+1,k))
+                        +beta_z*inv_dy*0.5*(phi_arr(i,j+1,k  )-phi_arr(i,j,k  )
+                                          + phi_arr(i,j+1,k+1)-phi_arr(i,j,k+1)));
                 },
                 [=] AMREX_GPU_DEVICE (int i, int j, int k) {
                     By_arr(i,j,k) += inv_c * (
-                        -beta_z*inv_dx*( phi_arr(i+1,j,k)-phi_arr(i,j,k) )
-                        +beta_x*inv_dz*( phi_arr(i,j,k+1)-phi_arr(i,j,k) ));
+                        -beta_z*inv_dx*0.5*(phi_arr(i+1,j,k  )-phi_arr(i,j,k  )
+                                          + phi_arr(i+1,j,k+1)-phi_arr(i,j,k+1))
+                        +beta_x*inv_dz*0.5*(phi_arr(i  ,j,k+1)-phi_arr(i  ,j,k)
+                                          + phi_arr(i+1,j,k+1)-phi_arr(i+1,j,k)));
                 },
                 [=] AMREX_GPU_DEVICE (int i, int j, int k) {
                     Bz_arr(i,j,k) += inv_c * (
-                        -beta_x*inv_dy*( phi_arr(i,j+1,k)-phi_arr(i,j,k) )
-                        +beta_y*inv_dx*( phi_arr(i+1,j,k)-phi_arr(i,j,k) ));
+                        -beta_x*inv_dy*0.5*(phi_arr(i  ,j+1,k)-phi_arr(i  ,j,k)
+                                          + phi_arr(i+1,j+1,k)-phi_arr(i+1,j,k))
+                        +beta_y*inv_dx*0.5*(phi_arr(i+1,j  ,k)-phi_arr(i,j  ,k)
+                                          + phi_arr(i+1,j+1,k)-phi_arr(i,j+1,k)));
                 }
             );
 #else
             amrex::ParallelFor( tbx, tby, tbz,
                 [=] AMREX_GPU_DEVICE (int i, int j, int k) {
                     Bx_arr(i,j,k) += inv_c * (
-                        -beta_y*inv_dz*( phi_arr(i,j+1,k)-phi_arr(i,j,k) ));
+                        -beta_y*inv_dz*0.5*(phi_arr(i  ,j+1,k)-phi_arr(i  ,j,k)
+                                          + phi_arr(i+1,j+1,k)-phi_arr(i+1,j,k)));
                 },
                 [=] AMREX_GPU_DEVICE (int i, int j, int k) {
                     By_arr(i,j,k) += inv_c * (
-                        -beta_z*inv_dx*( phi_arr(i+1,j,k)-phi_arr(i,j,k) )
-                        +beta_x*inv_dz*( phi_arr(i,j+1,k)-phi_arr(i,j,k) ));
+                        -beta_z*inv_dx*0.5*(phi_arr(i+1,j  ,k)-phi_arr(i,j  ,k)
+                                          + phi_arr(i+1,j+1,k)-phi_arr(i,j+1,k))
+                        +beta_x*inv_dz*0.5*(phi_arr(i  ,j+1,k)-phi_arr(i  ,j,k)
+                                          + phi_arr(i+1,j+1,k)-phi_arr(i+1,j,k)));
                 },
                 [=] AMREX_GPU_DEVICE (int i, int j, int k) {
                     Bz_arr(i,j,k) += inv_c * (
-                        +beta_y*inv_dx*( phi_arr(i+1,j,k)-phi_arr(i,j,k) ));
+                        +beta_y*inv_dx*0.5*(phi_arr(i+1,j  ,k)-phi_arr(i,j  ,k)
+                                          + phi_arr(i+1,j+1,k)-phi_arr(i,j+1,k)));
```

Comment:
It seems that the original 2D code for B is correct.

Reply:
Ah, yes, that's right - except for `By` which is cell-centered in both x and z. I'll update the PR.